core/src/main/java/edu/wpi/grip/core/http/GripServer.java

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    private boolean didStart = false;
+    private boolean stopped = false;
+
+    /**
+     * Starts this server. Has no effect if the server has already been started.
+     */
+    public void start() {
+        if (!didStart && !stopped) {
+            // Don't call server.start() if it's already been started or stopped
+            server.start();
+            didStart = true;
+            stopped = false;
+        }
+    }
+
+    /**
+     * Stops this server. Note that a shutdown hook has been registered to call
+     * this method, so it's unlikely that this should need to be called.
+     */
+    public void stop() {
+        if(!stopped) {
+            server.stop(0);
+            didStart = false;
+            stopped = true;
+        }
+    }
+    
+    
+    /**
+     * Restarts the server on the current port.
+     */
+    public void restart() {
+        try {
+            stop();
+            server = HttpServer.create(new InetSocketAddress("localhost", getPort()), BACKLOG);
+            start();
+        } catch (IOException | IllegalStateException ex) {
+            Logger.getLogger(GripServer.class.getName()).log(Level.SEVERE, null, ex);
+            throw new GripException("Could not restart GripServer", ex);
+        }
+    }
+    
+    /**
+     * Gets the port this server is running on.
+     */
+    public int getPort() {
+        return server.getAddress().getPort();
+    }
+
+    /**
+     * Parses request parameters from a URI. For example, the URI
+     * "http://roborio-190-frc.local:8080/GRIP/data?foo=1;bar=2" will be parsed to
+     * {"foo"="1", "bar"="2"}.
+     */
+    private static Map<String, List<String>> getRequestParameters(final URI requestUri) {
+        final Map<String, List<String>> requestParameters = new LinkedHashMap<>();
+        final String requestQuery = requestUri.getRawQuery();
+        if (requestQuery != null) {
+            final String[] rawRequestParameters = requestQuery.split("[&;]"); // [foo=1, bar=2, ...]
+            for (final String rawRequestParameter : rawRequestParameters) {
+                final String[] requestParameter = rawRequestParameter.split("=", 2); // [foo, 1]
+                final String requestParameterName = decodeUrlComponent(requestParameter[0]);
+                requestParameters.putIfAbsent(requestParameterName, new ArrayList<>());
+                final String requestParameterValue = requestParameter.length > 1 ? decodeUrlComponent(requestParameter[1]) : null;
+                requestParameters.get(requestParameterName).add(requestParameterValue);
+            }
+        }
+        return requestParameters;
+    }
+
+    private static String decodeUrlComponent(final String urlComponent) {
+        try {
+            return URLDecoder.decode(urlComponent, CHARSET.name());
+        } catch (final UnsupportedEncodingException ex) {
+            // Will never happen, UTF-8 is supported by all JVMs
+            throw new GripException(ex);
```

Comment:
[That's not the correct exception to use](https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalStateException.html).

Normally I'd stick a `@SneakyThrows(UnsupportedEncodingException.class)` on this method, but adding lombok as a dependency just for this would be overkill.